### ISSUE TITLE
[ppc64le/pytorch] test_milstm_fusion_cuda and  test_lstm_fusion_cuda fails

### ISSUE THREAD

**Issue by avmgithub** [pytorch/pytorch#10726]
The test using jit compiler seems ok for test_lstm_fusion_cuda  .  But not sure why the test_lstm_fusion_cuda (__main__.TestScript) fails. 

test_milstm_fusion_cuda fails (it has no equivalent jit test)
  
How can I correct this ?

Pytorch version :  torch (0.5.0a0+f0d8a36)
Cuda 9.2 
cudnn 7
ppc64le Power8

Here is the build/test log:  https://powerci.osuosl.org/job/pytorch-linux-cuda9-cudnn7-py3-mpi-build-test-gpu/264/console

``` 
======================================================================
FAIL: test_lstm_fusion_cuda (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/pytorch/test/common.py", line 251, in wrapper
    method(*args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 107, in wrapper
    fn(*args, **kwargs)
  File "test_jit.py", line 2591, in test_lstm_fusion_cuda
    self.assertExpectedGraph(backward_graph(module), subname='backward')
  File "test_jit.py", line 228, in assertExpectedGraph
    self.assertExpected(str(graph), *args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 514, in assertExpected
    self.assertMultiLineEqual(expected, s)
AssertionError: 'grap[908 chars]aph(%9 : Float(3, 20)\n      %19 : Float(3, 20[1722 chars]n}\n' != 'grap[908 chars]aph(%11 : Float(3, 20)\n      %21 : Float(3, 2[1723 chars]n}\n'
  graph(%0 : Float(3, 20!)
        %1 : Float(3, 20!)
        %2 : Dynamic
        %3 : Dynamic
        %4 : Dynamic
        %5 : Dynamic
        %6 : Dynamic
        %7 : Dynamic
        %8 : Dynamic
        %x : Float(3, 10)
        %hx : Float(3, 20)
        %cx : Float(3, 20)
        %12 : Float(10!, 80!)
        %13 : Float(20!, 80!)
        %ingate : Float(3, 20)
        %forgetgate : Float(3, 20)
        %cellgate : Float(3, 20)
        %outgate : Float(3, 20)
        %18 : Float(3, 20)) {
    %19 : int = prim::Constant[value=1]()
    %20 : Float(3, 80) = prim::FusionGroup_0[device=0](%ingate, %forgetgate, %cellgate, %outgate, %cx, %1, %18, %0)
    %21 : Float(3, 80) = aten::mul(%20, %19)
    %22 : Float(80!, 3!) = aten::t(%21)
    %23 : Float(80, 20) = aten::mm(%22, %hx)
    %24 : Float(80!, 3!) = aten::t(%20)
    %25 : Float(80, 10) = aten::mm(%24, %x)
    return (%25, %23, %21, %21);
  }
- with prim::FusionGroup_0 = graph(%9 : Float(3, 20)
?                                   ^
+ with prim::FusionGroup_0 = graph(%11 : Float(3, 20)
?                                   ^^
-       %19 : Float(3, 20)
?         -
+       %21 : Float(3, 20)
?        +
        %33 : Float(3, 20)
-       %39 : Float(3, 20)
?        ^^
+       %41 : Float(3, 20)
?        ^^
        %46 : Float(3, 20)
        %53 : Float(3, 20!)
        %65 : Float(3, 20)
        %67 : Float(3, 20!)) {
    %69 : Float(3, 20) = aten::mul(%67, %65)
-   %68 : Float(3, 20) = aten::mul(%67, %39)
?                                        ^^
+   %68 : Float(3, 20) = aten::mul(%67, %41)
?                                        ^^
    %66 : Float(3, 20) = aten::mul(%65, %65)
    %64 : Float(3, 20) = aten::neg(%66)
    %61 : int = prim::Constant[value=1]()
    %62 : Float(3, 20) = aten::add(%64, %61, %61)
    %59 : Float(3, 20) = aten::mul(%68, %62)
    %55 : int = prim::Constant[value=1]()
    %56 : Float(3, 20) = aten::add(%53, %59, %55)
    %51 : int = prim::Constant[value=1]()
    %52 : Float(3, 20) = aten::mul(%56, %51)
    %50 : Float(3, 20) = aten::mul(%52, %33)
-   %49 : Float(3, 20) = aten::mul(%52, %9)
?                                        ^
+   %49 : Float(3, 20) = aten::mul(%52, %11)
?                                        ^^
    %47 : Float(3, 20) = aten::mul(%56, %46)
+   %44 : Float(3, 20) = aten::mul(%69, %41)
-   %44 : Float(3, 20) = aten::neg(%39)
?     ^                             ^^
+   %42 : Float(3, 20) = aten::neg(%41)
?     ^                             ^^
-   %42 : int = prim::Constant[value=1]()
?    ^^
+   %39 : int = prim::Constant[value=1]()
?    ^^
-   %43 : Float(3, 20) = aten::add(%44, %42, %42)
-   %40 : Float(3, 20) = aten::mul(%69, %39)
?                              ^^^  ^
+   %40 : Float(3, 20) = aten::add(%42, %39, %39)
?                              ^^^  ^^^^^^
-   %37 : Float(3, 20) = aten::mul(%40, %43)
?                                    ^    ^
+   %37 : Float(3, 20) = aten::mul(%44, %40)
?                                    ^    ^
    %34 : Float(3, 20) = aten::mul(%33, %33)
    %32 : Float(3, 20) = aten::neg(%34)
    %29 : int = prim::Constant[value=1]()
    %30 : Float(3, 20) = aten::add(%32, %29, %29)
    %27 : Float(3, 20) = aten::mul(%49, %30)
-   %24 : Float(3, 20) = aten::neg(%19)
-   %22 : int = prim::Constant[value=1]()
-   %23 : Float(3, 20) = aten::add(%24, %22, %22)
-   %20 : Float(3, 20) = aten::mul(%47, %19)
?     ^                                   -
+   %24 : Float(3, 20) = aten::mul(%47, %21)
?     ^                                  +
-   %17 : Float(3, 20) = aten::mul(%20, %23)
-   %14 : Float(3, 20) = aten::neg(%9)
?    ^^                             ^
+   %22 : Float(3, 20) = aten::neg(%21)
?    ^^                             ^^
-   %12 : int = prim::Constant[value=1]()
?     ^
+   %19 : int = prim::Constant[value=1]()
?     ^
-   %13 : Float(3, 20) = aten::add(%14, %12, %12)
?    ^^                             ^^    ^    ^
+   %20 : Float(3, 20) = aten::add(%22, %19, %19)
?    ^^                             ^^    ^    ^
+   %17 : Float(3, 20) = aten::mul(%24, %20)
-   %10 : Float(3, 20) = aten::mul(%50, %9)
?     ^                                  ^
+   %14 : Float(3, 20) = aten::mul(%50, %11)
?     ^                                  ^^
+   %12 : Float(3, 20) = aten::neg(%11)
+   %9 : int = prim::Constant[value=1]()
+   %10 : Float(3, 20) = aten::add(%12, %9, %9)
-   %7 : Float(3, 20) = aten::mul(%10, %13)
?                                   ^    ^
+   %7 : Float(3, 20) = aten::mul(%14, %10)
?                                   ^    ^
    %4 : Float(3, 80) = prim::FusedConcat[dim=1](%7, %17, %27, %37)
    return (%4);
  }


======================================================================
FAIL: test_milstm_fusion_cuda (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/pytorch/test/common.py", line 251, in wrapper
    method(*args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 107, in wrapper
    fn(*args, **kwargs)
  File "test_jit.py", line 2603, in test_milstm_fusion_cuda
    self.assertExpectedGraph(backward_graph(module), subname='backward')
  File "test_jit.py", line 228, in assertExpectedGraph
    self.assertExpected(str(graph), *args, **kwargs)
  File "/home/jenkins/pytorch/test/common.py", line 514, in assertExpected
    self.assertMultiLineEqual(expected, s)
AssertionError: 'grap[1502 chars]aph(%9 : Float(3, 20)\n      %19 : Float(3, 20[2751 chars]n}\n' != 'grap[1502 chars]aph(%11 : Float(3, 20)\n      %21 : Float(3, 2[2752 chars]n}\n'
  graph(%0 : Float(3, 20!)
        %1 : Float(3, 20!)
        %2 : Dynamic
        %3 : Dynamic
        %4 : Dynamic
        %5 : Dynamic
        %6 : Dynamic
        %7 : Dynamic
        %8 : Dynamic
        %9 : Dynamic
        %10 : Dynamic
        %11 : Dynamic
        %x : Float(3, 10)
        %hx : Float(3, 20)
        %cx : Float(3, 20)
        %alpha : Float(80)
        %beta_i : Float(80)
        %beta_h : Float(80)
        %18 : Float(10!, 80!)
        %Wx : Float(3, 80)
        %20 : Float(20!, 80!)
        %Uz : Float(3, 80)
        %22 : Float(3, 80)
        %ingate : Float(3, 20)
        %forgetgate : Float(3, 20)
        %cellgate : Float(3, 20)
        %outgate : Float(3, 20)
        %27 : Float(3, 20)) {
    %28 : Float(3, 80) = prim::FusionGroup_0[device=0](%ingate, %forgetgate, %cellgate, %outgate, %cx, %1, %27, %0)
    %29 : int[] = prim::Constant[value=[3, 80]]()
    %30 : int = prim::Constant[value=0]()
    %31 : Float(3!, 80) = aten::expand(%beta_h, %29, %30)
    %32 : Float(3!, 80) = aten::expand(%beta_i, %29, %30)
    %33 : Float(3!, 80) = aten::expand(%alpha, %29, %30)
    %34 : Float(3, 80), %35 : Float(3, 80), %36 : Float(3, 80), %37 : Float(3, 80), %38 : Float(3, 80), %39 : Float(3, 80) = prim::FusionGroup_1[device=0](%33, %32, %Wx, %28, %Uz, %22, %31)
    %40 : Float(80!, 3!) = aten::t(%36)
    %41 : Float(80, 20) = aten::mm(%40, %hx)
    %42 : Float(80!, 3!) = aten::t(%34)
    %43 : Float(80, 10) = aten::mm(%42, %x)
    return (%43, %41, %35, %37, %38, %39);
  }
- with prim::FusionGroup_0 = graph(%9 : Float(3, 20)
?                                   ^
+ with prim::FusionGroup_0 = graph(%11 : Float(3, 20)
?                                   ^^
-       %19 : Float(3, 20)
?         -
+       %21 : Float(3, 20)
?        +
        %33 : Float(3, 20)
-       %39 : Float(3, 20)
?        ^^
+       %41 : Float(3, 20)
?        ^^
        %46 : Float(3, 20)
        %53 : Float(3, 20!)
        %65 : Float(3, 20)
        %67 : Float(3, 20!)) {
    %69 : Float(3, 20) = aten::mul(%67, %65)
-   %68 : Float(3, 20) = aten::mul(%67, %39)
?                                        ^^
+   %68 : Float(3, 20) = aten::mul(%67, %41)
?                                        ^^
    %66 : Float(3, 20) = aten::mul(%65, %65)
    %64 : Float(3, 20) = aten::neg(%66)
    %61 : int = prim::Constant[value=1]()
    %62 : Float(3, 20) = aten::add(%64, %61, %61)
    %59 : Float(3, 20) = aten::mul(%68, %62)
    %55 : int = prim::Constant[value=1]()
    %56 : Float(3, 20) = aten::add(%53, %59, %55)
    %51 : int = prim::Constant[value=1]()
    %52 : Float(3, 20) = aten::mul(%56, %51)
    %50 : Float(3, 20) = aten::mul(%52, %33)
-   %49 : Float(3, 20) = aten::mul(%52, %9)
?                                        ^
+   %49 : Float(3, 20) = aten::mul(%52, %11)
?                                        ^^
    %47 : Float(3, 20) = aten::mul(%56, %46)
+   %44 : Float(3, 20) = aten::mul(%69, %41)
-   %44 : Float(3, 20) = aten::neg(%39)
?     ^                             ^^
+   %42 : Float(3, 20) = aten::neg(%41)
?     ^                             ^^
-   %42 : int = prim::Constant[value=1]()
?    ^^
+   %39 : int = prim::Constant[value=1]()
?    ^^
-   %43 : Float(3, 20) = aten::add(%44, %42, %42)
-   %40 : Float(3, 20) = aten::mul(%69, %39)
?                              ^^^  ^
+   %40 : Float(3, 20) = aten::add(%42, %39, %39)
?                              ^^^  ^^^^^^
-   %37 : Float(3, 20) = aten::mul(%40, %43)
?                                    ^    ^
+   %37 : Float(3, 20) = aten::mul(%44, %40)
?                                    ^    ^
    %34 : Float(3, 20) = aten::mul(%33, %33)
    %32 : Float(3, 20) = aten::neg(%34)
    %29 : int = prim::Constant[value=1]()
    %30 : Float(3, 20) = aten::add(%32, %29, %29)
    %27 : Float(3, 20) = aten::mul(%49, %30)
-   %24 : Float(3, 20) = aten::neg(%19)
-   %22 : int = prim::Constant[value=1]()
-   %23 : Float(3, 20) = aten::add(%24, %22, %22)
-   %20 : Float(3, 20) = aten::mul(%47, %19)
?     ^                                   -
+   %24 : Float(3, 20) = aten::mul(%47, %21)
?     ^                                  +
-   %17 : Float(3, 20) = aten::mul(%20, %23)
-   %14 : Float(3, 20) = aten::neg(%9)
?    ^^                             ^
+   %22 : Float(3, 20) = aten::neg(%21)
?    ^^                             ^^
-   %12 : int = prim::Constant[value=1]()
?     ^
+   %19 : int = prim::Constant[value=1]()
?     ^
-   %13 : Float(3, 20) = aten::add(%14, %12, %12)
?    ^^                             ^^    ^    ^
+   %20 : Float(3, 20) = aten::add(%22, %19, %19)
?    ^^                             ^^    ^    ^
+   %17 : Float(3, 20) = aten::mul(%24, %20)
-   %10 : Float(3, 20) = aten::mul(%50, %9)
?     ^                                  ^
+   %14 : Float(3, 20) = aten::mul(%50, %11)
?     ^                                  ^^
+   %12 : Float(3, 20) = aten::neg(%11)
+   %9 : int = prim::Constant[value=1]()
+   %10 : Float(3, 20) = aten::add(%12, %9, %9)
-   %7 : Float(3, 20) = aten::mul(%10, %13)
?                                   ^    ^
+   %7 : Float(3, 20) = aten::mul(%14, %10)
?                                   ^    ^
    %4 : Float(3, 80) = prim::FusedConcat[dim=1](%7, %17, %27, %37)
    return (%4);
  }
  with prim::FusionGroup_1 = graph(%5 : Float(3!, 80)
        %8 : Float(3!, 80)
        %10 : Float(3, 80)
        %12 : Float(3, 80)
        %13 : Float(3, 80)
        %20 : Float(3, 80)
        %22 : Float(3!, 80)) {
    %30 : int = prim::Constant[value=1]()
    %29 : int = prim::Constant[value=1]()
    %28 : int = prim::Constant[value=1]()
    %26 : int = prim::Constant[value=1]()
    %27 : Float(3, 80) = aten::mul(%12, %26)
    %25 : Float(3, 80) = aten::mul(%27, %13)
    %24 : Float(3, 80) = aten::mul(%27, %10)
    %23 : Float(3, 80) = aten::mul(%27, %22)
    %21 : Float(3, 80) = aten::mul(%12, %20)
    %19 : int = prim::Constant[value=1]()
    %17 : int = prim::Constant[value=1]()
    %18 : Float(3, 80) = aten::add(%23, %21, %17)
    %14 : Float(3, 80) = aten::mul(%12, %13)
    %11 : Float(3, 80) = aten::mul(%14, %10)
    %9 : Float(3, 80) = aten::mul(%27, %8)
    %6 : Float(3, 80) = aten::mul(%14, %5)
    %2 : int = prim::Constant[value=1]()
    %3 : Float(3, 80) = aten::add(%9, %6, %2)
    return (%3, %11, %18, %24, %25, %27);
  }

**Comment by avmgithub** [comment]
@zou3519 when you get the chance , can you please comment on this.  I see you've added these tests in the last few days or so.   Is this due to an architecture difference ?

**Comment by zou3519** [comment]
@avmgithub as far as I know this should not be architecture dependent. It looks like the changes might just be differences in variable naming (maybe that is not deterministic?) 

It's a little hard for me to read a the output as a diff, it would really help me if you could run the following on your machine
```
python test/test_jit.py TestScript.test_lstm_fusion_cuda --accept 
```
(this will modify test/expect/TestScript.test_lstm_fusion_cuda-backward.expect)
and then upload `test/expect/TestScript.test_lstm_fusion_cuda-backward.expect`.

Thank you!

**Comment by avmgithub** [comment]
@zou3519 here is the test_lstm_fusion_cuda-backward.expect file
[TestScript.test_lstm_fusion_cuda-backward.expect.txt](https://github.com/pytorch/pytorch/files/2307684/TestScript.test_lstm_fusion_cuda-backward.expect.txt)

**Comment by avmgithub** [comment]
@ezyang here is the TestScript.test_milstm_fusion_cuda-backward.expect file 

[TestScript.test_milstm_fusion_cuda-backward.expect.txt](https://github.com/pytorch/pytorch/files/2307703/TestScript.test_milstm_fusion_cuda-backward.expect.txt)

**Comment by avmgithub** [comment]
@zou3519 just wondering if you had a chance to look at this.

**Comment by zou3519** [comment]
Closed by #11579 I believe
